### PR TITLE
The great toxification (part 4)

### PR DESCRIPTION
**Pull request by stephenfin**
Subject: Clean up requirements

Clean up requirements (`test-reqs.txt`, `setup.py`) to ensure virtualenv and non-virtualenv builds are identical

### Feature or Bugfix
- Feature

### Purpose
To paraphrase the Zen of Python (import this):

> tox is one honking great idea -- let's use more of it!

Start modernizing the test infrastructure that's using make and other hand-crafted tools by replacing them with tox-based derivatives.

**This is part four.**

### Detail
This is a multi-step process, with the end goal of completely removing the Makefile and running everything through `tox`. The first fours steps of these can be done in parallel, but combined they block the final two.

1. Make `tox` a little more friendly by enabling extras like coverage, test profiling etc. by default, and closing gaps with the current Makefile targets such as PyLint and building docs in particular formats
2. Replace/remove custom tooling found in `utils` in favor of modern alternatives like `flake8`
3. Centralize as much configuration as possible inside `setup.cfg` and `tox.ini`, to avoid polluting the top level directory.
4. Clean up requirements (`test-reqs.txt`, `setup.py`) to ensure virtualenv and non-virtualenv builds are identical (**this change**)

The remaining two steps are blocked by the above:

5. Switch CIs to use `tox` instead of the Makefile, and update `CONTRIBUTING.rst` to refer to tox. Deprecate the Makefile
6. (Bonus) Integrate [codecov.io](https://codecov.io) to make use of the coverage stats we now have

### Relates
- N/A

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -63,25 +63,27 @@
     ':python_version<"3.5"': [
         'typing'
     ],
+    ':python_version<"3.4"': [
+        'enum34'
```
Comment (tk0miya) [inline setup.py]: Hmm? Why did you add this to dependency?

File: setup.py [hunk 2]
```diff
@@ -63,25 +63,27 @@
...
+    ],
     'websupport': [
         'sqlalchemy>=0.9',
         'whoosh>=2.0',
     ],
     'test': [
         'pytest',
-        'mock',  # it would be better for 'test:python_version in 2.7'
-        'simplejson',  # better: 'test:platform_python_implementation=="PyPy"'
```
Comment (tk0miya) [inline setup.py:73]: +1; now pypy provides optimized json package :-)

**Review by tk0miya — CHANGES_REQUESTED** [review]
Basically LGTM, but it seems all tests are failed. could you check it please?

File: setup.py (hunk 1 above)
Comment (stephenfin) [reply setup.py]: It's used in the Sphinx source, though currently surrounded by a `try-except ImportError`. I figured we should just make it a requirement so we can remove that `try-except`. Did I misinterpret something?

Comment (tk0miya) [reply setup.py]: Yes, it's not dependency. It is used to inspect user's program on autodoc module when the program uses enum. So it is not requirements of Sphinx itself.

Comment (stephenfin) [reply setup.py]: OK, I'll add it to `tests` instead so

Comment (stephenfin) [reply setup.py]: Wait - we distribute `autodoc` as part of Sphinx. Therefore shouldn't we provide all the dependencies for that module with Sphinx?

Comment (tk0miya) [reply setup.py]: `autodoc` uses `enum` only if it is installed. It means `autodoc` imports it only if the target program uses it. So `enum34` is a requirement of target program.
In other words, it is not "requirements", but also "optional".

Comment (stephenfin) [reply setup.py]: OK, I'll move this to `test` so

**Comment by stephenfin** [comment]
> Basically LGTM, but it seems all tests are failed. could you check it please?

This is done now

File: .travis.yml [hunk 3]
```diff
@@ -40,9 +40,8 @@ addons:
 install:
   - pip install -U pip setuptools
   - pip install docutils==$DOCUTILS
-  - pip install -r test-reqs.txt
-  - if [[ $TRAVIS_PYTHON_VERSION == '3.6' ]]; then python3.6 -m pip install mypy typed-ast; fi
+  - pip install .[test,websupport]
+  - if [[ $TRAVIS_PYTHON_VERSION == '3.6' ]]; then python3.6 -m pip install mypy typed-ast flake8; fi
```
Comment (tk0miya) [inline .travis.yml]: Is it okay to execute flake8 only with python3.6? Once I saw a blog article saying flake8 should be executed with all interpreters.  Please let me know if you know.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
@shimizukawa I need your help for this. Could you review this please?

File: .travis.yml (hunk 3 above)
Comment (stephenfin) [reply .travis.yml]: This is a good point, but I don't think it's something to worry about too much for now. Part 5 will introduce `tox-travis` so we'll have a chance to completely rework what jobs we run and maybe use `flake8` on Py2.7 too

File: setup.py [hunk 4]
```diff
@@ -69,19 +69,19 @@
...
+        'typed_ast',
+    ],
 }
 
-# for sdist installation with pip-1.5.6
-if sys.platform == 'win32':
-    requires.append('colorama>=0.3.5')
-
-if sys.version_info < (3, 5):
-    requires.append('typing')
```
Comment (shimizukawa) [inline setup.py:83]: I think this removing lines means 'drop support pip-1.5.6'.
Maybe such environment is very rare nowadays, however it would be better if it is indicated on CHANGES.

**Review by shimizukawa — APPROVED** [review]
LGTM with nits

File: setup.py (hunk 4 above)
Comment (stephenfin) [reply setup.py:83]: @shimizukawa I've refrained from touching CHANGES because it causes merge conflicts and you guys usually do it. Should I do this or can I simply call it out in the commit message as I've done and leave the CHANGES update to maintainers?

On a side note, have you seen [`reno`](https://docs.openstack.org/reno/latest/)? Is this something you might be open to using in Sphinx in a future release?

File: .travis.yml (hunk 3 above)
Comment (stephenfin) [reply .travis.yml]: I've modified this now so it'll build against all versions of Python

**Comment by stephenfin** [comment]
I've resolved the merge conflict. In the process, I realized that there's actually no reason whatsoever to keep `test-req.txt` around so I've removed it. We weren't really using it for [it's true purpose](https://caremad.io/posts/2013/07/setup-vs-requirement/) so getting rid of it seemed like the thing to do.

**Comment by stephenfin** [comment]
@tk0miya @shimizukawa Is there anything more I need to do for this?

File: setup.py (hunk 4 above)
Comment (tk0miya) [reply setup.py:83]: Okay, I'll update CHANGES on merging.

**Comment by tk0miya** [comment]
It seems all review are passed now. So I'll merge this tonight (in JST)

**Comment by tk0miya** [comment]
Merged! Thank you for your contribution!

**Comment by stephenfin** [comment]
Lovely. Thanks for the reviews, @tk0miya :smile:
